curve_test.go

```diff
@@ -146,3 +152,216 @@ func TestMultAir(t *testing.T) {
+		{
+			k:         StrToBig("2"),
+			expectedX: StrToBig("3324833730090626974525872402899302150520188025637965566623476530814354734325"),
+			expectedY: StrToBig("3147007486456030910661996439995670279305852583596209647900952752170983517249"),
+		},
+		{
+			k:         StrToBig("3"),
+			expectedX: StrToBig("1839793652349538280924927302501143912227271479439798783640887258675143576352"),
+			expectedY: StrToBig("3564972295958783757568195431080951091358810058262272733141798511604612925062"),
+		},
+	}
+
+	for _, tt := range testMult {
+		x, y, err := Curve.PrivateToPoint(tt.k)
+
+		if err != nil {
+			t.Errorf("EcMult %v\n", err)
+		}
+
+		if x.Cmp(tt.expectedX) != 0 {
+			t.Errorf("ResX %v does not == expected %v\n", x, tt.expectedX)
+
+		}
+		if y.Cmp(tt.expectedY) != 0 {
+			t.Errorf("ResY %v does not == expected %v\n", y, tt.expectedY)
+		}
+	}
+}
+
+func BenchmarkEcMultAll(b *testing.B) {
+	var _genBigIntBits = func(n int) (k *big.Int) {
+		k = big.NewInt(1)
+		for i := 1; i < n; i++ {
+			k = k.Lsh(k, 1).Add(k, big.NewInt(1))
+		}
+		return
+	}
+
+	testEcMult := []struct {
+		algo string
+		fn   EcMultiFn
+	}{
+		{
+			algo: "Add-Always-Double",
+			fn:   Curve.EcMult, // used algo
+		},
+		{
+			algo: "Double-And-Add",
+			fn:   Curve.ecMult_DoubleAndAdd, // original algo
+		},
+		{
+			algo: "Montgomery-Ladder",
+			fn:   Curve.ecMult_Montgomery,
+		},
+		{
+			algo: "Montgomery-Ladder-Lsh",
+			fn:   Curve.ecMult_MontgomeryLsh,
+		},
+	}
+
+	var out strings.Builder
+	for _, tt := range testEcMult {
+		// test (+ time) injected ec multi fn performance via Curve.privateToPoint
+		var _test = func(k *big.Int) int64 {
+			start := time.Now()
+			Curve.privateToPoint(k, tt.fn)
+			return time.Since(start).Nanoseconds()
+		}
+
+		xs := []float64{}
+		// generate numbers with 1 to 251 bits set
+		for i := 1; i < Curve.N.BitLen(); i++ {
+			k := _genBigIntBits(i)
+			b.Run(fmt.Sprintf("%s/input_bits_len/%d", tt.algo, k.BitLen()), func(b *testing.B) {
+				ns := _test(k)
+				xs = append(xs, float64(ns))
+			})
+		}
+
+		// generate numbers with 1 to 250 trailing zero bits set
+		k := _genBigIntBits(Curve.N.BitLen() - 1)
+		for i := 1; i < Curve.N.BitLen()-1; i++ {
+			k.Rsh(k, uint(i)).Lsh(k, uint(i))
+			b.Run(fmt.Sprintf("%s/input_bits_len/%d#%d", tt.algo, k.BitLen(), k.TrailingZeroBits()), func(b *testing.B) {
+				ns := _test(k)
+				xs = append(xs, float64(ns))
+			})
+		}
+
+		// computes the weighted mean of the dataset.
+		// we don't have any weights (ie: all weights are 1) so we pass a nil slice.
+		mean := stat.Mean(xs, nil)
+		variance := stat.Variance(xs, nil)
+		stddev := math.Sqrt(variance)
+
+		out.WriteString("-----------------------------\n")
+		out.WriteString(fmt.Sprintf("algo=       %v\n", tt.algo))
+		out.WriteString(fmt.Sprintf("stats(ns)\n"))
+		out.WriteString(fmt.Sprintf("  mean=     %v\n", mean))
+		out.WriteString(fmt.Sprintf("  variance= %v\n", variance))
+		out.WriteString(fmt.Sprintf("  std-dev=  %v\n", stddev))
```

Comment:
We could verify that expected algorithm (Add-Always-Double) is the one with min std-dev